### DISCUSSION TITLE
ui.div()

### DISCUSSION THREAD

**Discussion by ofenbach** [zauberzeug/nicegui#572] [Ideas / Feature Requests]
Why not add this shortcut instead of using `ui.element('div')`?

its easy to implement i guess?

**Comment by ofenbach** [comment]
```python
def div():
    return ui.element('div')
```
    
this is how i implemented it now, for example for this usage:

```python
with div():
    ...
```

**Comment by falkoschindler** [comment]
Interesting idea, @ofenbach! Let me share my thoughts about it:

1. We try to decouple NiceGUI from historic HTML terminology to make it more accessible to non-web developers. Therefore one of our most primitive elements is called "label" instead of "div", even though it _is_ a div.
2. Sure, adding a `ui.div` would be very easy to implement. But we try not to spoil the namespace with too many elements. Once they are in the world, we have to support them basically forever. So we're getting a bit picky. 😉
3. We could make the tag name optional, so that `ui.element()` creates a div by default. It would, however, introduce the danger of forgetting to specify a tag name when deriving from `Element`. This is not a major problem, but something to consider.
4. What are you using `div()` for? We usually use `ui.row` and `ui.column` as a container for multiple elements. If they don't suit your needs, creating a custom class (or function) might just be the right thing to do.

**Reply by ofenbach** [reply]
I use div quite a lot to style my responsive components.
Since cat design uses a lot of responsive mobile design layouts with flexbox and will feature a grid element like MUI
and other custom elements that are not available in default NiceGUI i use it quite often.
But i implemented the div now for my purposes.

Just realized that my code base is becoming more like a fork (or level on top of nicegui) than a design system :)

The design system so far is highly customizable as well. 

Will give you examples once the code base is ready to be discussed with you experts from zauberzeug 👍

**Reply by ofenbach** [reply]
the catdesign got updatet, if you got time you can check it out.
i created a landing page to see all the elements available.
its not really responsive but if you look at the code inside cat_design.py you can see i made
custom div and box elements in case you are interested.

**Reply by rodja** [reply]
This is the url of the CatDesign project: https://github.com/ofenbach/CatDesign. I'll hopefully check it out tomorrow.

**Reply by rodja** [reply]
Looks really great. Can't wait for an official pip package :-)
We plan on setting up a collection of cool projects done with NiceGUI in #396. This is definitely one of them!

Maybe your design system could benefit from auto-completion with literals as was discussed in https://github.com/zauberzeug/nicegui/discussions/117#discussioncomment-4869309?

**Reply by ofenbach** [reply]
Thanks for the feedback. 

That's exactly what i was looking for (Literals). Will implement those.

**Reply by falkoschindler** [reply]
In https://github.com/zauberzeug/nicegui/pull/362 we're working on something related. Should be part of the next release.

**Comment by dridk** [comment]
I understand your reasoning but I think nicegui should have both low-level and high-level usage.
I think it would have been better to do what JustPy did by separating the basic components ( ui.div ) from the quasar components ( ui.QButton ). 

For example, I prefer to use ui.element(‘div’) rather than ui.cards() because I can fully customize the element with tailwindcss. If I use ui.cards(), I always have style conflicts between props and classes.

I'm currently trying to reproduce this design: https://ui.shadcn.com/. I'll let you know if I succeed

**Reply by falkoschindler** [reply]
Your post sparked the idea for a separate `html` namespace which could be auto-generated from some kind of HTML specification and include all HTML elements and potentially their attributes:
```py
from nicegui import ui, html

with html.p():
    html.img(src="...")
    html.a(href="...")
    html.text("This is inner HTML of the p tag.")

ui.run()
```
All these statements would translate to a `ui.element` with the corresponding tag, which can be further manipulated with style, classes, props and event registrations.

What do you think?

**Reply by dridk** [reply]
Looks good but with classes and style only. I don't understand the purpose of "props" when we can use function arguments.

**Reply by dridk** [reply]
Just to share my thoughts, In a more generalist perspective, I would have done :

```python
from niceui.quasar import ui
from niceui import html 
ui.button()
html.button()
```

But I know it's too complicated because it would require a lot of work.  is already an incredible libs you hav done.

**Comment by falkoschindler** [comment]
In version [2.5.0](https://github.com/zauberzeug/nicegui/releases/tag/v2.5.0) we introduced an html module including `html.div()`.
Therefore I'll close this feature request as completed.
